docs/DesignDocs/libc_initialization.md

```diff
@@ -0,0 +1,169 @@
+# Musl libc Initialization
+
+## Normal musl libc initialization on Linux x86\_64
+
+A normal executable linux program with uses Musl libc will have the following initialization pattern,
+where `_start` is the entry point of every program.
+
+
+```
+_start()
+    v
+    |
+    --> _start_c()
+              v
+              |
+              --> _libc_start_main()
+                            v
+                            |
+                            --> __init_libc()
+                            --> _libc_start_main_stage2()
+                                          v
+                                          |
+                                          --> __libc_start_init()
+                                          --> main()
+                                          --> exit()
+```
+
+### What do each of these functions do?
+#### \_start
+
+`_start` is the entry point for every userspace executable. Its only
+purpose is to initialize rbp to 0 and call `_start_c`.
+
+#### \_start\_c
+
+Sets some registers and calls `_libc_start_main`
+
+#### \_libc\_start\_main
+
+Calls `__init_libc` and `_libc_start_main_stage2`
+
+#### \_\_init\_libc
+
+Finally something interesting. `__init_libc` does many things:
+
+* Calls `__init_tls` to initialize thread local storage
+* Calls `__init_ssp` to initialize stack protector
+* Calls poll syscall to check if stdin, stdout, and stderr to be ready for I/O. If any are not available for IO, it initializes them to /dev/null
+* sets the following global variables:
+    * \_\_environ
+    * \_\_hwcap
+    * \_\_sysinfo
+    * \_\_progname
+* Sets the following members of the global libc structure
+    * auxv
+    * page\_size
+    * secure
+
+#### \_libc\_start\_main\_stage2
+
+Calls `__libc_start_init`. Then calls `main` to do final initialization.
+Then calls `exit` with the return code of `main` when main exits.
+
+#### \_\_libc\_start\_init:
+
+Calls `_init` and loops through `__init_array` and calls each of the
+compiler-generated init functions.
+
+## OE Initialization
+
+### OE libc Initialization
+
+OE initializes libc in similar ways, but leaves out some of the initialization
+that is not needed in the enclave. Additional musl-specific initialization is
+done by a function `_oe_init_c` which is marked with the constructor attribute.
+This ensures that enclaves which link libc will call `_oe_init_c` during
+oe_call_init_functions.
+
+#### \_\_init\_tls
+
+OE initializes thread local storage via a call to `oe_thread_local_init` in
+enclave/core/linux/threadlocal.c which is called during `td_init()` which
+initializes a thread data structure.
+
+#### \_\_init\_ssp
+
+OE initializes the stack protector again during the `td_init()` function.
+
+#### Check for availability of stdin, stdout, and stderr
+
+OE enclaves do console IO (and all file IO as of this writing) by making
```

Comment:
I think the "and all file IO as of this writing" is likely to get out of date without noticing that this file and line needs to be updated.   The feature.new_platforms branch does have a file IO provider that does not make ocalls to the host, I don't know what the status is of porting it to integrate with musl libc and merging it into master, but the point is we know that this claim here is short-lived and easy to miss in the future.  Hence would be better to avoid making this claim here, especially since the section isn't about file IO per se but is just about stdin/stdout/stderr.